### PR TITLE
feature/new-model-config

### PR DESCRIPTION
**Pull request by joaogsleite**
Refer to issue #48 for more information

### CODE REVIEW

**Comment by joaogsleite** [comment]
* graphql pagination
  * added `offset` and `limit` args to every graphql table query
  * created `tableCount` query for each table to return total items count
* removed all frontend logic (chips, filter, input types, etc...)
* created a `visible` field on each column config to specify if field should be returned in **list**, **edit** or **relation** requests 
* moved all frontend related configs to `layout` field inside **table** and **columns** configs
* updated config generator
* fixed all table rest controllers for new config structure
* updated tests for new config structure

**Comment by JWebCoder** [comment]
@joaogsleite please add tests to the new Graphql count resolver

File: src/generator/configurationTypes.ts [hunk 1]
```diff
@@ -57,76 +57,51 @@ export interface IManyToOneRelation {
...
     [key in Hooks]?: {
       before?: IHookFunction,
       after?: IHookFunction,
     }
   },
+  layout?: any,
+}
+
+export interface ITableRelation {
+  type: '1:n' | 'n:1' | 'm:n',
```
Comment (JWebCoder) [inline src/generator/configurationTypes.ts:79]: Please differentiate the type of relations using multiple typings.
I will ease up the development and creation of future configurations.

File: src/generator/mysql/index.ts [hunk 2]
```diff
@@ -114,14 +114,14 @@ function createConnection() {
...
 const getDatabaseData = (answers: any): PromiseLike<IDatabaseData[]> => {
-  console.log(answers)
   userAnswers = answers
   return new Promise<any>(
     (res, rej) => {
       const connection = createConnection()
       connection.execute(
         'show tables',
         (err, results: mysql.RowDataPacket[]) => {
+          console.log('results', results)
```
Comment (JWebCoder) [inline src/generator/mysql/index.ts]: remove the console.log

File: src/generator/configurationTypes.ts [hunk 3]
```diff
@@ -57,76 +57,51 @@ export interface IManyToOneRelation {
...
+  remoteForeignKey?: string,
+  remoteTable: string,
 }
 
 export interface IColumnRelation {
-  type: string,
+  type: 'n:1',
   table: string,
   key: string,
-  display: string,
```
Comment (JWebCoder) [inline src/generator/configurationTypes.ts:115]: Where can we set the column to display in a relational field on the new configuration? I would like to keep being able to define the display field without the need to touch the frontend.

File: test/configs/models/products.js [hunk 4]
```diff
@@ -1,155 +1,204 @@
...
+      "model": {
+        "type": "smallint",
+        "allowNull": true
       },
-      "editable": true,
-      "input": {
-        "type": "number"
+      "layout": {
+        "label": "Type",
+        "listColumn": {},
```
Comment (JWebCoder) [inline test/configs/models/products.js:94]: the default value on the config should be a boolean

File: src/api/controllers/TablesController.ts [hunk 5]
```diff
@@ -26,11 +26,13 @@ class TablesController {
...
         }
         if (isAuthorized) {
-          return {
-            name: table.name,
-            verbose: table.verbose,
-            order: table.order || undefined,
-          }
+          const result: any = {}
+          Object.keys(table).forEach((key) => {
+            if (!['columns', 'relations'].includes(key)) {
```
Comment (JWebCoder) [inline src/api/controllers/TablesController.ts]: This is way to slow, please do a simple check.

File: src/api/controllers/TablesController.ts [hunk 6]
```diff
@@ -26,11 +26,13 @@ class TablesController {
...
+              result[key] = (table as any)[key]
```
Comment (JWebCoder) [inline src/api/controllers/TablesController.ts:32]: Please check this a little bit more to see if it's possible to remove the as any.

File: src/api/controllers/TablesController.ts [hunk 7]
```diff
@@ -26,11 +26,13 @@ class TablesController {
...
+            }
+          })
+          return result
         }
         return undefined
```
Comment (JWebCoder) [inline src/api/controllers/TablesController.ts]: should be only "return"

File: src/api/graphql/typeBuilder.ts [hunk 8]
```diff
@@ -36,13 +39,23 @@ const types: {
 } = {}
 
 export function buildFields(table: ITableInfo, options: IBuildTypeOptions = { relations: true } ) {
```
Comment (JWebCoder) [inline src/api/graphql/typeBuilder.ts]: please make the pagination true by default.

File: src/generator/configurationTypes.ts [hunk 9]
```diff
@@ -57,76 +57,51 @@ export interface IManyToOneRelation {
...
+    list: boolean,
     detail: boolean,
+    relation: boolean,
   },
-  editable: boolean,
-  input: {
-    type: 'text' | 'checkbox' | 'date' | 'number',
+  model: {
+    isPk?: boolean,
+    type: string,
```
Comment (JWebCoder) [inline src/generator/configurationTypes.ts]: the string type should be an enum.

**Review by JWebCoder — CHANGES_REQUESTED** [review]
Please, consider the comments. We need to get the typings right.

File: src/generator/configurationTypes.ts (hunk 3 above)
Comment (joaogsleite) [reply src/generator/configurationTypes.ts:115]: We can set column.visible.relation = true to force the backend to return that column when the table is included as a relation. On the layout object, we can set which field is used to display the item "title".

File: src/api/graphql/typeBuilder.ts (hunk 8 above)
Comment (joaogsleite) [reply src/api/graphql/typeBuilder.ts]: I've changed the pagination to true by default but most don't use pagination:
Screenshot from 2020-01-24 15-20-52

File: test/configs/models/products.js (hunk 4 above)
Comment (JWebCoder) [reply test/configs/models/products.js:94]: This is to be applied to all fields that take an object or a boolean has an input.

**Review by JWebCoder — APPROVED** [review]

**Pull request merged**